src/commons/editor/Editor.tsx

```diff
@@ -50,529 +64,249 @@ type StateProps = {
-
-    // TODO: Removal
-    /* disable error threshold incrementer
-
-    const jshintOptions = {
-      // undef: true,
-      // unused: true,
-      esnext: true,
-      moz: true,
-      devel: true,
-      browser: true,
-      node: true,
-      laxcomma: true,
-      laxbreak: true,
-      lastsemic: true,
-      onevar: false,
-      passfail: false,
-      maxerr: 1000,
-      expr: true,
-      multistr: true,
-      globalstrict: true
-    };
-    session.$worker.send('setOptions', [jshintOptions]);
-
-    */
-
-    // NOTE: the two `any`s below are because the Ace editor typedefs are
-    // hopelessly incomplete
-    editor.on('gutterclick' as any, this.handleGutterClick);
-
-    // Change all info annotations to error annotations
-    session.on('changeAnnotation' as any, this.handleAnnotationChange(session));
-
-    // Start autocompletion
-    acequire('ace/ext/language_tools').setCompleters([this.completer]);
+const getModeString = (chapter: number, variant: Variant, library: string) =>
+  'source' + chapter.toString() + variant + library;
 
-    // Has session ID
-    if (this.props.editorSessionId !== '') {
-      this.handleStartCollabEditing(editor);
-    }
+/**
+ * This _modifies global state_ and defines a new Ace mode globally.
+ *
+ * Don't call this directly in render functions!
+ */
+const selectMode = (chapter: number, variant: Variant, library: string) => {
+  HighlightRulesSelector(chapter, variant, library, Documentation.externalLibraries[library]);
+  ModeSelector(chapter, variant, library);
+};
 
-    this.handleVariableHighlighting();
+const makeHandleGutterClick = (
+  handleEditorUpdateBreakpoints: DispatchProps['handleEditorUpdateBreakpoints']
+) => (e: AceMouseEvent) => {
+  const target = e.domEvent.target! as HTMLDivElement;
+  if (
+    target.className.indexOf('ace_gutter-cell') === -1 ||
+    !e.editor.isFocused() ||
+    e.clientX > 35 + target.getBoundingClientRect().left
+  ) {
+    return;
   }
 
-  public componentWillUnmount() {
-    if (this.ShareAce !== null) {
-      // Umounting, closing websocket
-      this.ShareAce.WS.close();
-    }
-    this.ShareAce = null;
+  // Breakpoint related.
+  const row = e.getDocumentPosition().row;
+  const content = e.editor.session.getLine(row);
+  const breakpoints = e.editor.session.getBreakpoints();
+  if (
+    breakpoints[row] === undefined &&
+    content.length !== 0 &&
+    !content.includes('//') &&
+    !content.includes('debugger;')
+  ) {
+    e.editor.session.setBreakpoint(row, undefined!);
+  } else {
+    e.editor.session.clearBreakpoint(row);
   }
+  e.stop();
+  handleEditorUpdateBreakpoints(e.editor.session.getBreakpoints());
+};
 
-  public componentDidUpdate(prevProps: EditorProps) {
-    const newCursorPosition = this.props.newCursorPosition;
-    if (newCursorPosition && newCursorPosition !== prevProps.newCursorPosition) {
-      this.moveCursor(newCursorPosition);
+const makeHandleAnnotationChange = (session: any) => () => {
```

Comment:
We should type `session` better - I've dug around and it seems this should be typed as `Ace.EditSession`.

Reply:
Addressed in b5a249b.